presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java

```diff
@@ -186,38 +187,37 @@ public Page getOutput()
     @VisibleForTesting
     static DecodedBlockNode decodeBlock(BlockFlattener flattener, Block block)
     {
-        try (BlockLease lease = flattener.flatten(block)) {
-            Block decodedBlock = lease.get();
+        BlockLease lease = flattener.flatten(block);
```

Comment:
In a separate PR https://github.com/prestodb/presto/pull/13455, @mbasmanova suggested a different fix:

Keep track of all the leases produced by the flattener and close them when no longer needed. Here is how this might look like:

```
        private final BlockFlattener flattener = new BlockFlattener();
        private final Closer blockLeaseCloser = Closer.create();
...
            for (int i = 0; i < decodedBlocks.length; i++) {
                decodedBlocks[i] = decodeBlock(flattener, blockLeaseCloser, page.getBlock(i));
            }
...
    static DecodedBlockNode decodeBlock(BlockFlattener flattener, Closer blockLeaseCloser, Block block)
    {
        BlockLease lease = flattener.flatten(block);
        blockLeaseCloser.register(lease::close);
        Block decodedBlock = lease.get();

        if (decodedBlock instanceof ArrayBlock) {
            ColumnarArray columnarArray = ColumnarArray.toColumnarArray(decodedBlock);
            return new DecodedBlockNode(columnarArray, ImmutableList.of(decodeBlock(flattener, blockLeaseCloser, columnarArray.getElementsBlock())));
        }

        if (decodedBlock instanceof MapBlock) {
            ColumnarMap columnarMap = ColumnarMap.toColumnarMap(decodedBlock);
            return new DecodedBlockNode(columnarMap, ImmutableList.of(decodeBlock(flattener, blockLeaseCloser, columnarMap.getKeysBlock()), decodeBlock(flattener, blockLeaseCloser, columnarMap.getValuesBlock())));
        }

        if (decodedBlock instanceof RowBlock) {
            ColumnarRow columnarRow = ColumnarRow.toColumnarRow(decodedBlock);
            ImmutableList.Builder<DecodedBlockNode> children = ImmutableList.builder();
            for (int i = 0; i < columnarRow.getFieldCount(); i++) {
                children.add(decodeBlock(flattener, blockLeaseCloser, columnarRow.getField(i)));
            }
            return new DecodedBlockNode(columnarRow, children.build());
        }

        if (decodedBlock instanceof DictionaryBlock) {
            return new DecodedBlockNode(decodedBlock, ImmutableList.of(decodeBlock(flattener, blockLeaseCloser, ((DictionaryBlock) decodedBlock).getDictionary())));
        }

        if (decodedBlock instanceof RunLengthEncodedBlock) {
            return new DecodedBlockNode(decodedBlock, ImmutableList.of(decodeBlock(flattener, blockLeaseCloser, ((RunLengthEncodedBlock) decodedBlock).getValue())));
        }

        return new DecodedBlockNode(decodedBlock, ImmutableList.of());
    }
...
            // Close block leases
            try {
                blockLeaseCloser.close();
            }
            catch (IOException e) {
                throw new UncheckedIOException(e);
```
This change is in 32ef717604 Fix "invalid position in block" error now.